### PR TITLE
MM-17023 - Integrations Marketplace

### PR DESCRIPTION
**Pull request by marianunez**
#### Summary

- Added Plugins Marketplace full-screen modal that will display all plugins available from the Marketplace Server. 
- Plugins Marketplace can be accessed from the Main Menu only available for admin users with general `MANAGE_SYSTEM` permissions and if both `Enable Marketplace` and `Enable Plugins` settings are turned on.
- Initial version includes the ability to:
   - Install plugins within the Marketplace dialog
   - Link to plugin configuration screen in System Console.

Screen Shot 2019-08-29 at 8 57 34 PM

#### Ticket Link

[Marketplace Epic](https://mattermost.atlassian.net/browse/MM-17023)
[Invision Design](https://projects.invisionapp.com/share/QESMWLJBSTZ#/screens)

#### Related Pull Requests

- Server changes in the Integrations Marketplace [feature branch](https://github.com/mattermost/mattermost-server/tree/mm-17023-integrations-marketplace)
- Redux changes [PR here](https://github.com/mattermost/mattermost-redux/pull/912)

### CODE REVIEW

File: components/integrations_marketplace/marketplace_modal.jsx [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+            />);
+        if (installedList) {
+            noPluginsMessage = (
+                <FormattedMessage
+                    id='marketplace_modal.no_plugins_installed'
+                    defaultMessage='You do not have any plugins installed.'
+                />);
+        }
+
+        return (<div className='more-modal__list'>
```
Comment (jespino) [inline components/integrations_marketplace/marketplace_modal.jsx]: I think we can simplify it with a small bit of repetition. I would replace the whole block with something like:

```js
if (pluginsArray.length === 0) {
    let noPluginsMessage = ...
    return <div...
}
return <div...>{pluginsArray.map(p) => {...}}</div>
```

This block is a bit hard to follow and with a bit extra structure can be more simple.

File: components/integrations_marketplace/marketplace_modal.jsx [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+                >
+                    <div className='MarketplaceModal'>
+                        <h1>
+                            <FormattedMessage
+                                id='marketplace_modal.title'
+                                defaultMessage='Plugins Marketplace'
+                            />
+                        </h1>
+                        {input}
+                        <Tabs
```
Comment (jespino) [inline components/integrations_marketplace/marketplace_modal.jsx]: Not a big fan of using react-bootstrap stuff for new things, but because we don't have yet a proper alternative to that, I think is the correct way to do it :)

File: components/integrations_marketplace/marketplace_modal.jsx [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+                                {this.state.loading ?
+                                    <LoadingScreen/> : (
+                                        this.getPluginsListContent(this.props.marketplacePlugins, false)
+                                    ) }
+                            </Tab>
+                            <Tab
+                                eventKey={MarketplaceTabs.INSTALLED_PLUGINS}
+                                title={localizeMessage('marketplace_modal.tabs.installed_plugins', `Installed (${this.props.installedPlugins.length})`)}
+                            >
+                                {this.getPluginsListContent(this.props.installedPlugins, false)}
```
Comment (jespino) [inline components/integrations_marketplace/marketplace_modal.jsx]: shouldn't be here the second parameter `true`?

File: components/integrations_marketplace/marketplace_modal.scss [hunk 4]
```diff
@@ -0,0 +1,119 @@
+@import 'utils/variables';
+
+.MarketplaceModal {
```
Comment (jespino) [inline components/integrations_marketplace/marketplace_modal.scss]: Love see new components using the new approach :heart:

File: components/svg/plugin_icon.jsx [hunk 5]
```diff
@@ -0,0 +1,52 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
```
Comment (jespino) [inline components/svg/plugin_icon.jsx]: Do you mind to move it to `components/widgets/icons/`?

File: components/integrations_marketplace/marketplace_modal.jsx (hunk 2 above)
Comment (marianunez) [reply components/integrations_marketplace/marketplace_modal.jsx]: Possible ticket for creating our own Tab widget?

File: components/integrations_marketplace/marketplace_modal.jsx (hunk 3 above)
Comment (marianunez) [reply components/integrations_marketplace/marketplace_modal.jsx]: Yes! 🤦‍♀️Good catch!

File: components/integrations_marketplace/marketplace_modal.jsx (hunk 2 above)
Comment (jespino) [reply components/integrations_marketplace/marketplace_modal.jsx]: Yes, that could be great. Probably the first version can use the bootstrap tab under the hood, and if we want to get rid of react-boostrap we only have to get rid of the component in the widget, not everywhere.

Comment (jespino) [reply components/integrations_marketplace/marketplace_modal.jsx]: @reflog you are working in the list of reusable components, makes sense to add the "Tabs/Tab" widgets?

File: components/integrations_marketplace/marketplace_modal.jsx (hunk 1 above)
Comment (marianunez) [reply components/integrations_marketplace/marketplace_modal.jsx]: Done. Much better! Thanks

File: components/svg/plugin_icon.jsx (hunk 5 above)
Comment (marianunez) [reply components/svg/plugin_icon.jsx]: Done

File: components/integrations_marketplace/marketplace_item/marketplace_item.jsx [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+
+            this.state = {
+                installed: this.props.installed,
+                installing: false,
+                confirmOverwriteInstallModal: false,
+                serverError: null,
+            };
+        }
+
+        onClickOvewriteModal = () => {
```
Comment (lieut-data) [inline components/integrations_marketplace/marketplace_item/marketplace_item.jsx]: Do we still need a confirmation flow for overwriting (aka upgrading, I presume)?

File: components/integrations_marketplace/marketplace_modal.jsx (hunk 2 above)
Comment (reflog) [reply components/integrations_marketplace/marketplace_modal.jsx]: @jespino - yep, it will be part of the list!

File: components/integrations_marketplace/index.js [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+        installedPlugins: getMarketplaceInstalledPlugins(state),
+        marketplacePlugins: state.entities.plugins.marketplacePlugins,
+    };
+}
+
+function mapDispatchToProps(dispatch) {
+    return {
+        actions: bindActionCreators({
+            closeModal: () => closeModal(ModalIdentifiers.INTEGRATIONS_MARKETPLACE),
+            installPluginFromUrl,
```
Comment (ali-farooq0) [inline components/integrations_marketplace/index.js]: should this be removed? doesn't look like it's being used.

File: components/integrations_marketplace/marketplace_modal.jsx [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+            loading: true,
+        };
+    }
+
+    componentWillMount() {
+        this.getMarketplacePlugins();
+    }
+
+    getMarketplacePlugins = async () => {
+        await this.props.actions.getMarketplacePlugins();
```
Comment (ali-farooq0) [inline components/integrations_marketplace/marketplace_modal.jsx]: what'll happen if this call fails, not sure what is the pattern for catching errors, is that part of the mm-redux action?

Comment (marianunez) [reply components/integrations_marketplace/marketplace_modal.jsx]: Correct, the error handling will be done in redux with a specific action that will have error information. WIP

File: components/integrations_marketplace/marketplace_item/marketplace_item.jsx [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+        onInstall = () => {
+            this.setState({installing: true});
+            this.installPlugin(false);
+        }
+
+        render() {
+            const ariaLabel = `${this.props.name}, ${this.props.description}`.toLowerCase();
+            const versionLabel = `(${this.props.version})`;
+
+            let button = (
```
Comment (ali-farooq0) [inline components/integrations_marketplace/marketplace_item/marketplace_item.jsx]: would it be worth turning this into its own internal component, that resides in this file and is not exported?

File: components/integrations_marketplace/marketplace_item/marketplace_item.jsx [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+            actions: PropTypes.shape({
+                installPluginFromUrl: PropTypes.func.isRequired,
+            }).isRequired,
+        };
+
+        constructor(props) {
+            super(props);
+
+            this.state = {
+                installed: this.props.installed,
```
Comment (ali-farooq0) [inline components/integrations_marketplace/marketplace_item/marketplace_item.jsx]: doesn't look like this `installed` field is set anything, worth just using `this.props.installed` all the time?

Comment (marianunez) [reply components/integrations_marketplace/marketplace_item/marketplace_item.jsx]: This is passed in to the local state because if its installed it changes it locally to update the button to `Configure`. It's being used here: https://github.com/mattermost/mattermost-webapp/blob/bde6b01e23ddd5bc8a9292991b850fb4c067eba5/components/integrations_marketplace/marketplace_item/marketplace_item.jsx#L81
Pushing a missing change I have locally that missed the last commit that sets `installed` instead.

File: components/integrations_marketplace/marketplace_modal.jsx [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+                            onSelect={this.handleSelect}
+                            unmountOnExit={true}
+                        >
+                            <Tab
+                                eventKey={MarketplaceTabs.ALL_PLUGINS}
+                                title={localizeMessage('marketplace_modal.tabs.all_plugins', 'All Plugins')}
+                            >
+                                {this.state.loading ?
+                                    <LoadingScreen/> : (
+                                        this.getPluginsListContent(this.props.marketplacePlugins, false)
```
Comment (ali-farooq0) [inline components/integrations_marketplace/marketplace_modal.jsx]: This definitely works, but wondering if we can break this down into smaller components. The overall architecture doesn't need to change, this just turns into a more component styled structure.

I was imagining we could do something like:
```
{this.state.loading ? <LoadingScreen/> : (<PluginsTab type={MarketplaceTabs.ALL_PLUGINS />)}

...
{this.state.loading ? <LoadingScreen/> : (<PluginsTab  type={MarketplaceTabs.INSTALLED_PLUGINS />)}
```


```
PluginsTab extends React.Component {
    render() {
        return (
            <div>
                {
                    this.props.list.length == 0 &&
                    <EmptyPlugins
                        text={
                            this.props.type == INSTALLED
                            ? <FormattedMessage
                                id='marketplace_modal.no_plugins_installed'
                                defaultMessage='You do not have any plugins installed.'
                            />)
                            : <FormattedMessage
                                id='marketplace_modal.no_plugins'
                                defaultMessage='There are no plugins available at this time.'
                            />)
                        }
                    >
                    {
                        this.props.type == INSTALLED &&
                        <a onClick={() => this.changeTab(MarketplaceTabs.ALL_PLUGINS)}>
                            <FormattedMessage
                                id='marketplace_modal.install_plugins'
                                defaultMessage='Install Plugins'
                            />
                        </a>
                    }
                    </EmptyPlugins>
                }
                {
                    pluginsArray.map((p) => {
                        return (
                            <MarketplaceItem
                                key={p.Manifest.id}
                                id={p.Manifest.id}
                                name={p.Manifest.name}
                                description={p.Manifest.description}
                                version={p.Manifest.version}
                                isPrepackaged={false}
                                downloadUrl={p.DownloadURL}
                                homepageUrl={p.HomepageURL}
                                installed={p.InstalledVersion !== ''}
                                onConfigure={this.close}
                                onInstalled={this.getMarketplacePlugins}
                            />
                        );
                    })
                }
            </div>
        )
    }
}

EmptyPlugins extends React.Component {
    render() {
        return (
            <div className='no_plugins_div'>
                <br/>
                <PluginIcon className='icon__plugin'/>
                <br/>
                <br/>
                {this.props.text}
                <br/>
                <br/>
                {this.props.children}
            </div>
        );
    }
}
```

Not blocking this review for this refactor, but would be kinda nice if did do it :) 2/5 on this. Using declarative components increases maintainability and reusability.
Thoughts?

File: components/integrations_marketplace/marketplace_modal.jsx [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+                    <FormattedMessage
+                        id='marketplace_modal.no_plugins_installed'
+                        defaultMessage='You do not have any plugins installed.'
+                    />);
+            }
+
+            return (<div className='no_plugins_div'>
+                <br/>
+                <PluginIcon className='icon__plugin'/>
+                <br/>
```
Comment (ali-farooq0) [inline components/integrations_marketplace/marketplace_modal.jsx]: if we're using `<br/>` for padding, we should probably use scss padding for this. 1/5 on this :)

https://stackoverflow.com/questions/1726073/is-it-sometimes-bad-to-use-br

File: components/integrations_marketplace/marketplace_modal.scss [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+    }
+
+    h2 {
+        font-weight: 300;
+        strong {
+            font-weight: 400;
+        }
+    }
+
+    .search_input {
```
Comment (ali-farooq0) [inline components/integrations_marketplace/marketplace_modal.scss]: minor nit: `.search-input` or `.search__input`?

File: components/integrations_marketplace/marketplace_item/marketplace_item.jsx (hunk 10 above)
Comment (marianunez) [reply components/integrations_marketplace/marketplace_item/marketplace_item.jsx]: https://github.com/mattermost/mattermost-webapp/pull/3492/commits/8d46b8ceba0236e592ffbf70d6fda20fd1aae734

File: components/integrations_marketplace/marketplace_modal.scss [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+        margin-right: 4px;
+
+        svg {
+            fill: var(--button-bg-60);
+            width: 32px;
+            height: 32px;
+        }
+    }
+
+    .no_plugins_div{
```
Comment (ali-farooq0) [inline components/integrations_marketplace/marketplace_modal.scss]: nit; `-` or `_`?

File: components/integrations_marketplace/marketplace_modal.scss [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+        text-align: center
+    }
+
+    .item_error {
```
Comment (ali-farooq0) [inline components/integrations_marketplace/marketplace_modal.scss]: same as above

File: components/integrations_marketplace/marketplace_modal.scss [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+        background-color: var(--error-text-08);
+    }
+
+    .error_text {
```
Comment (ali-farooq0) [inline components/integrations_marketplace/marketplace_modal.scss]: same as above

File: components/integrations_marketplace/index.js (hunk 7 above)
Comment (marianunez) [reply components/integrations_marketplace/index.js]: Good catch 👍Leftover from refactoring

File: components/main_menu/main_menu.jsx [hunk 17]
```diff
@@ -13,6 +13,8 @@ import {useSafeUrl} from 'utils/url';
 import * as UserAgent from 'utils/user_agent.jsx';
 import InvitationModal from 'components/invitation_modal';
 
+import MarketplaceModal from 'components/integrations_marketplace';
```
Comment (ali-farooq0) [inline components/main_menu/main_menu.jsx]: should this be moved to the bottom on line 23, with the rest of modals?

File: components/integrations_marketplace/marketplace_modal.scss (hunk 13 above)
Comment (marianunez) [reply components/integrations_marketplace/marketplace_modal.scss]: This was the conventions that was used on other `.scss` files in the webapp. @asaadmahmood maybe has more insight on what convention we should follow here.

File: components/integrations_marketplace/marketplace_modal.scss (hunk 15 above)
Comment (marianunez) [reply components/integrations_marketplace/marketplace_modal.scss]: @asaadmahmood thoughts?

File: components/integrations_marketplace/marketplace_modal.jsx (hunk 12 above)
Comment (marianunez) [reply components/integrations_marketplace/marketplace_modal.jsx]: I was battling with the css to get it just like @asaadmahmood design and resorted to <br/>🙈

Maybe @asaadmahmood can help on this too.

File: components/integrations_marketplace/marketplace_modal.jsx [hunk 18]
```diff
@@ -0,0 +1,189 @@
...
+                            unmountOnExit={true}
+                        >
+                            <Tab
+                                eventKey={MarketplaceTabs.ALL_PLUGINS}
+                                title={localizeMessage('marketplace_modal.tabs.all_plugins', 'All Plugins')}
+                            >
+                                {this.state.loading ?
+                                    <LoadingScreen/> : (
+                                        this.getPluginsListContent(this.props.marketplacePlugins, false)
+                                    ) }
```
[the remark this answers is not in the document]
Comment (marianunez) [reply components/integrations_marketplace/marketplace_modal.jsx]: Sure. Lint didn't pick it up 🤔I've noticed it has a hard time with the `?` shorthand if.

File: components/main_menu/main_menu.jsx (hunk 17 above)
Comment (marianunez) [reply components/main_menu/main_menu.jsx]: Good linting 🙃Done!

Comment (ali-farooq0) [reply components/main_menu/main_menu.jsx]: 🤓

File: components/integrations_marketplace/marketplace_item/marketplace_item.jsx (hunk 9 above)
Comment (marianunez) [reply components/integrations_marketplace/marketplace_item/marketplace_item.jsx]: Not sure, it handles several props and local state that would need to be passed down and IMO not much gained on punting in it's own internal component.

Comment (ali-farooq0) [reply components/integrations_marketplace/marketplace_item/marketplace_item.jsx]: fair, can we at least move it out of the render function? and turn it into it's own?

File: components/integrations_marketplace/marketplace_modal.jsx (hunk 11 above)
Comment (marianunez) [reply components/integrations_marketplace/marketplace_modal.jsx]: Agreed that this would be a good improvement. I will wrap up pending error handling case and will come back to this

Comment (ali-farooq0) [reply components/integrations_marketplace/marketplace_modal.jsx]: 👍

File: components/integrations_marketplace/marketplace_item/marketplace_item.jsx (hunk 9 above)
Comment (marianunez) [reply components/integrations_marketplace/marketplace_item/marketplace_item.jsx]: Done!

File: components/integrations_marketplace/marketplace_modal.jsx [hunk 19]
```diff
@@ -141,12 +143,24 @@ export default class MarketplaceModal extends React.Component {
             </div>
         );
 
+        let errorBanner = null;
+        if (this.props.serverError) {
+            errorBanner = (
+                <div
+                    className='error-bar'
+                    dangerouslySetInnerHTML={{__html: Markdown.format(localizeMessage(this.props.serverError.server_error_id,
+                        'Error connecting to the marketplace server.  Please check your settings in the [System Console](/admin_console/plugins/plugin_management).'))}}
```
Comment (ali-farooq0) [inline components/integrations_marketplace/marketplace_modal.jsx]: has double spacing between `server.` and `Please check`

File: components/integrations_marketplace/marketplace_modal.jsx [hunk 20]
```diff
@@ -0,0 +1,204 @@
...
+                        inputComponent={LocalizedInput}
+                        onInput={this.doSearch}
+                    />
+                </div>
+            </div>
+        );
+
+        let errorBanner = null;
+        if (this.props.serverError) {
+            errorBanner = (
```
Comment (ali-farooq0) [inline components/integrations_marketplace/marketplace_modal.jsx]: really surprising that we don't have an error banner, thought we did though, maybe with a different styling?

Comment (marianunez) [reply components/integrations_marketplace/marketplace_modal.jsx]: There is an announcement banner but styling is different, specially because it's styled to be used in the top of the whole app. For marketplace, it was needed in the full-screen modal so this `div` was a simple solution for our needs.

File: components/integrations_marketplace/marketplace_modal.jsx (hunk 19 above)
Comment (marianunez) [reply components/integrations_marketplace/marketplace_modal.jsx]: Fixed
